{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a88a15b",
   "metadata": {},
   "source": [
    "### File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f03ee35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "rna_seq_df  = pd.read_csv(\"data/RNA_Proteomics_RNAlevel.csv\")\n",
    "prot_seq_df = pd.read_csv(\"data/RNA_Proteomics_Proteinlevel.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef3fe5ce",
   "metadata": {},
   "source": [
    "### 1. Find CAI for RNA level file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a15966eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from math import log, exp\n",
    "\n",
    "# Kazusa Codon usage \n",
    "# https://www.kazusa.or.jp/codon/cgi-bin/showcodon.cgi?species=9606\n",
    "kazusa_freq = {\n",
    "    \"UUU\": 17.6, \"UUC\": 20.3, \"UUA\": 7.7, \"UUG\": 12.9,\n",
    "    \"CUU\": 13.2, \"CUC\": 19.6, \"CUA\": 7.2, \"CUG\": 39.6,\n",
    "    \"AUU\": 16.0, \"AUC\": 20.8, \"AUA\": 7.5, \"AUG\": 22.0,\n",
    "    \"GUU\": 11.0, \"GUC\": 14.5, \"GUA\": 7.1, \"GUG\": 28.1,\n",
    "    \"UCU\": 15.2, \"UCC\": 17.7, \"UCA\": 12.2, \"UCG\": 4.4,\n",
    "    \"CCU\": 17.5, \"CCC\": 19.8, \"CCA\": 16.9, \"CCG\": 6.9,\n",
    "    \"ACU\": 13.1, \"ACC\": 18.9, \"ACA\": 15.1, \"ACG\": 6.1,\n",
    "    \"GCU\": 18.4, \"GCC\": 27.7, \"GCA\": 15.8, \"GCG\": 7.4,\n",
    "    \"UAU\": 12.2, \"UAC\": 15.3, \"UAA\": 1.0, \"UAG\": 0.8,\n",
    "    \"CAU\": 10.9, \"CAC\": 15.1, \"CAA\": 12.3, \"CAG\": 34.2,\n",
    "    \"AAU\": 17.0, \"AAC\": 19.1, \"AAA\": 24.4, \"AAG\": 31.9,\n",
    "    \"GAU\": 21.8, \"GAC\": 25.1, \"GAA\": 29.0, \"GAG\": 39.6,\n",
    "    \"UGU\": 10.6, \"UGC\": 12.6, \"UGA\": 1.6, \"UGG\": 13.2,\n",
    "    \"CGU\": 4.5, \"CGC\": 10.4, \"CGA\": 6.2, \"CGG\": 11.4,\n",
    "    \"AGU\": 12.1, \"AGC\": 19.5, \"AGA\": 12.2, \"AGG\": 12.0,\n",
    "    \"GGU\": 10.8, \"GGC\": 22.2, \"GGA\": 16.5, \"GGG\": 16.5\n",
    "}\n",
    "\n",
    "aa_table = {\n",
    "    'F': ['UUU', 'UUC'], \n",
    "    'L': ['UUA','UUG','CUU','CUC','CUA','CUG'],\n",
    "    'I': ['AUU','AUC','AUA'], 'M': ['AUG'], \n",
    "    'V': ['GUU','GUC','GUA','GUG'],\n",
    "    'S': ['UCU','UCC','UCA','UCG','AGU','AGC'], \n",
    "    'P': ['CCU','CCC','CCA','CCG'],\n",
    "    'T': ['ACU','ACC','ACA','ACG'], \n",
    "    'A': ['GCU','GCC','GCA','GCG'],\n",
    "    'Y': ['UAU','UAC'], \n",
    "    'H': ['CAU','CAC'], \n",
    "    'Q': ['CAA','CAG'],\n",
    "    'N': ['AAU','AAC'], \n",
    "    'K': ['AAA','AAG'], \n",
    "    'D': ['GAU','GAC'],\n",
    "    'E': ['GAA','GAG'], \n",
    "    'C': ['UGU','UGC'], \n",
    "    'W': ['UGG'],\n",
    "    'R': ['CGU','CGC','CGA','CGG','AGA','AGG'], \n",
    "    'G': ['GGU','GGC','GGA','GGG']\n",
    "}\n",
    "\n",
    "# Compute relative codon weights\n",
    "codon_weights = {}\n",
    "for aa, codons in aa_table.items():\n",
    "    mean_freq = sum(kazusa_freq[c] for c in codons if c in kazusa_freq) / len(codons)\n",
    "    rscu = {c: kazusa_freq[c] / mean_freq for c in codons if c in kazusa_freq}\n",
    "    max_rscu = max(rscu.values())\n",
    "    for c in rscu:\n",
    "        codon_weights[c] = rscu[c] / max_rscu\n",
    "\n",
    "\n",
    "# to compute CAI\n",
    "def compute_cai(seq):\n",
    "    if pd.isna(seq): \n",
    "        return None\n",
    "    seq = seq.upper().replace(\"T\",\"U\")\n",
    "    codons = [seq[i:i+3] for i in range(0, len(seq), 3) if len(seq[i:i+3])==3]\n",
    "    codons = [c for c in codons if c not in [\"UAA\",\"UAG\",\"UGA\"]]  # remove stop\n",
    "    ws = [codon_weights[c] for c in codons if c in codon_weights]\n",
    "    if not ws: \n",
    "        return None\n",
    "    return exp(sum(log(w) for w in ws) / len(ws))\n",
    "\n",
    "\n",
    "rna_seq_df[\"CAI\"] = rna_seq_df[\"CDS\"].apply(compute_cai)\n",
    "rna_seq_df.to_csv(\"data/RNA_Proteomics_RNAlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8c38722",
   "metadata": {},
   "source": [
    "### 2. CAI for the proteomics level file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8fe97c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_seq_df[\"CAI\"] = prot_seq_df[\"CDS\"].apply(compute_cai)\n",
    "prot_seq_df.to_csv(\"data/RNA_Proteomics_Proteinlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8bc161c",
   "metadata": {},
   "source": [
    "### 3. CAI for all proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da987245",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for record in SeqIO.parse(\"sequence.txt\", \"fasta\"):\n",
    "    nm_id = record.id.split(\"|\")[1] if \"|\" in record.id else record.id\n",
    "    seq = str(record.seq)\n",
    "    cai = compute_cai(seq)\n",
    "    results.append({\"NM_ID\": nm_id, \"CAI\": cai})\n",
    "\n",
    "cai_df = pd.DataFrame(results)\n",
    "cai_df.to_csv(\"data/sequence_CAI.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e922bcbc",
   "metadata": {},
   "source": [
    "### 4. CAI for codon97"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2887f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_df = pd.read_csv(\"data/RNA_Proteomics_Filtered.csv\")\n",
    "rna_df[\"CAI\"] = rna_df[\"CDS\"].apply(compute_cai)\n",
    "\n",
    "rna_df.to_csv(\"data/RNA_Proteomics_Filtered.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af7b05d8",
   "metadata": {},
   "source": [
    "### 5. Distribution for log2FC>1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0735ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rnalevel_file = \"data/RNA_Proteomics_RNAlevel.csv\"\n",
    "RNA_file = pd.read_csv(rnalevel_file)\n",
    "\n",
    "filtered = RNA_file[\n",
    "    (RNA_file[\"log2FoldChange\"] > 1) &\n",
    "    (RNA_file[\"padj\"] < 0.05)\n",
    "]\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_RNAlevel_LogFC>1.csv\"\n",
    "filtered.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d1943f",
   "metadata": {},
   "outputs": [],
   "source": [
    "protlevel_file = \"data/RNA_Proteomics_Proteinlevel.csv\"\n",
    "prot_file = pd.read_csv(protlevel_file)\n",
    "\n",
    "filtered = prot_file[\n",
    "    (prot_file[\"Log2FC\"] > 1) &\n",
    "    (prot_file[\"Welch's T-test q-value ZFC3H1 KO\"] < 0.05)\n",
    "]\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_Protlevel_LogFC>1.csv\"\n",
    "filtered.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2963db4e",
   "metadata": {},
   "source": [
    "### 6. Violin Plot for CAI distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59c98161",
   "metadata": {},
   "source": [
    "#### Bootstrap for the mean value (compare with all_prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac4a967",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def bootstrap_pvalue(test_vals, all_vals, n_iter=1000, seed=123):\n",
    "    np.random.seed(seed)\n",
    "    test_mean = np.mean(test_vals)\n",
    "    n = len(test_vals)\n",
    "\n",
    "    boot_means = []\n",
    "    for _ in range(n_iter):\n",
    "        sample = np.random.choice(all_vals, size=n, replace=False)\n",
    "        boot_means.append(np.mean(sample))\n",
    "\n",
    "    boot_means = np.array(boot_means)\n",
    "\n",
    "    pval = np.mean(np.abs(boot_means - np.mean(all_vals)) >= np.abs(test_mean - np.mean(all_vals)))\n",
    "\n",
    "    return test_mean, np.mean(all_vals), pval\n",
    "\n",
    "results = {}\n",
    "for label, values in {\n",
    "    \"Codon97\": codon97_cai,\n",
    "    \"Proteomics_log2FC>0\": protein_cai,\n",
    "    \"RNAseq_log2FC>0\": rna_cai,\n",
    "    \"RNAseq_Log2fc>1\": rna_log2FC1_cai,\n",
    "    \"Prot_Log2fc>1\": prot_log2FC1_cai\n",
    "}.items():\n",
    "    results[label] = bootstrap_pvalue(values, all_cai, n_iter=1000)\n",
    "\n",
    "res_df = pd.DataFrame(results, index=[\"Mean_Group\",\"Mean_Allprot\",\"p-value\"]).T\n",
    "print(res_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2ff17d",
   "metadata": {},
   "source": [
    "#### Violin Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de090120",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Files\n",
    "codon97_df = pd.read_csv(\"data/RNA_Proteomics_Filtered.csv\")\n",
    "protein_df = pd.read_csv(\"data/RNA_Proteomics_Proteinlevel.csv\")\n",
    "rna_df     = pd.read_csv(\"data/RNA_Proteomics_RNAlevel.csv\")\n",
    "all_df = pd.read_csv(\"data/sequence_CAI.csv\")\n",
    "rna_log2FC1_df = pd.read_csv(\"data/RNA_Proteomics_RNAlevel_LogFC>1.csv\")\n",
    "prot_log2FC1_df = pd.read_csv(\"data/RNA_Proteomics_Protlevel_LogFC>1.csv\")\n",
    "\n",
    "# Extract CAI coln in each source\n",
    "codon97_cai = codon97_df[\"CAI\"].dropna().tolist()\n",
    "protein_cai = protein_df[\"CAI\"].dropna().tolist()\n",
    "rna_cai     = rna_df[\"CAI\"].dropna().tolist()\n",
    "all_cai = all_df[\"CAI\"].dropna().tolist()\n",
    "rna_log2FC1_cai = rna_log2FC1_df[\"CAI\"].dropna().tolist()\n",
    "prot_log2FC1_cai = prot_log2FC1_df[\"CAI\"].dropna().tolist()\n",
    "\n",
    "# Put into one dataframe\n",
    "all_data = pd.DataFrame({\n",
    "    \"CAI\": codon97_cai + protein_cai + rna_cai +all_cai+rna_log2FC1_cai+prot_log2FC1_cai,\n",
    "    \"Source\": ([\"Codon97\"] * len(codon97_cai) +\n",
    "               [\"Proteomics_Log2FC>0\"] * len(protein_cai) +\n",
    "               [\"RNAseq_Log2FC>0\"] * len(rna_cai)+\n",
    "               [\"Allprot_all_Log2FC\"] * len(all_cai)+\n",
    "               [\"RNAseq_Log2fc>1\"]* len(rna_log2FC1_cai)+\n",
    "               [\"Prot_Log2fc>1\"]*len(prot_log2FC1_cai)) \n",
    "               \n",
    "})\n",
    "\n",
    "order = [\"Codon97\", \"Proteomics_Log2FC>0\", \"RNAseq_Log2FC>0\", \"RNAseq_Log2fc>1\", \"Prot_Log2fc>1\", \"Allprot_all_Log2FC\"]\n",
    "counts = all_data.groupby(\"Source\").size()\n",
    "\n",
    "# Plot the violin plot\n",
    "plt.figure(figsize=(12,6))\n",
    "\n",
    "ax = sns.violinplot(x=\"Source\", y=\"CAI\", data=all_data, \n",
    "                    inner=None, palette=\"Set3\", order=order,linewidth=1.2\n",
    "                    )\n",
    "\n",
    "# Add the jitter plot\n",
    "sns.stripplot(x=\"Source\", y=\"CAI\", data=all_data,\n",
    "              color=\"black\", alpha=0.2, jitter=0.25, size=1.5, order=order)\n",
    "\n",
    "# Add the box plot\n",
    "sns.boxplot(\n",
    "    x=\"Source\", y=\"CAI\", data=all_data,\n",
    "    width=0.15, showcaps=True, order=order,\n",
    "    boxprops={'facecolor':'grey', 'edgecolor':'black'},\n",
    "    whiskerprops={'color':'black'},\n",
    "    medianprops={'color':'black', 'linewidth':1},\n",
    "    showfliers=False \n",
    ")\n",
    "\n",
    "# Add group means\n",
    "group_means = all_data.groupby(\"Source\")[\"CAI\"].mean()\n",
    "for i, src in enumerate(order):\n",
    "    ax.scatter(i, group_means[src], color=\"white\", zorder=3, s=10, label=\"Mean\" if i==0 else \"\")\n",
    "\n",
    "# Add p-values\n",
    "p_values = {\n",
    "    \"Codon97\": 0.006,\n",
    "    \"Proteomics_Log2FC>0\": 0.000,\n",
    "    \"RNAseq_Log2FC>0\": 0.704,\n",
    "    \"RNAseq_Log2fc>1\": 0.121,\n",
    "    \"Prot_Log2fc>1\": 0.970\n",
    "}\n",
    "y_max = all_data[\"CAI\"].max()\n",
    "y_offset = 0.03  \n",
    "for i, src in enumerate(order[:-1]):  \n",
    "    x1, x2 = i, len(order)-1 \n",
    "    pval = p_values.get(src, None)\n",
    "    if pval is None: \n",
    "        continue\n",
    "    y = y_max + (i+1)*y_offset\n",
    "    plt.plot([x1, x1, x2, x2], [y, y+0.005, y+0.005, y], lw=1.2, c='k')\n",
    "    plt.text((x1+x2)/2, y+0.01, f\"p = {pval:.3f}\", ha='center', va='bottom', fontsize=10, color=\"blue\")\n",
    "\n",
    "plt.xticks(\n",
    "    ticks=range(len(order)),\n",
    "    labels=[f\"{src}\\n(n={counts[src]})\" for src in order]\n",
    ")\n",
    "# Y range\n",
    "plt.yticks(np.arange(0.6, 1.01, 0.05))  \n",
    "\n",
    "plt.ylim(all_data[\"CAI\"].min()-0.1, all_data[\"CAI\"].max()+0.2)\n",
    "plt.title(\"Distribution of CAI Across Sources\", fontsize=14)\n",
    "plt.ylabel(\"CAI\", fontsize=12)\n",
    "plt.xlabel(\"\")\n",
    "plt.grid(axis=\"y\", linestyle=\"--\", alpha=0.6)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save to svg format\n",
    "import matplotlib as mpl\n",
    "from pathlib import Path\n",
    "\n",
    "mpl.rcParams['svg.fonttype'] = 'none'\n",
    "\n",
    "outdir = Path(\"data/figs\")\n",
    "outdir.mkdir(parents=True, exist_ok=True)\n",
    "plt.savefig(outdir / \"CAI_violin.svg\", format=\"svg\", bbox_inches=\"tight\", facecolor=\"white\")\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
