{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "053147a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "codon97_df = pd.read_csv(\"/mnt/work_3/sijin/CAI/RNA_Proteomics_Filtered.csv\")\n",
    "protein_df = pd.read_csv(\"/mnt/work_3/sijin/CAI/RNA_Proteomics_Proteinlevel.csv\")\n",
    "rna_df     = pd.read_csv(\"/mnt/work_3/sijin/CAI/RNA_Proteomics_RNAlevel.csv\")\n",
    "\n",
    "def get_len(seq):\n",
    "    if pd.isna(seq):\n",
    "        return None\n",
    "    return len(str(seq).replace(\" \", \"\").replace(\"\\n\", \"\").upper())\n",
    "\n",
    "codon97_df[\"CDS_len\"] = codon97_df[\"CDS\"].apply(get_len)\n",
    "protein_df[\"CDS_len\"] = protein_df[\"CDS\"].apply(get_len)\n",
    "rna_df[\"CDS_len\"]     = rna_df[\"CDS\"].apply(get_len)\n",
    "\n",
    "# pull the fasta\n",
    "records = []\n",
    "for rec in SeqIO.parse(\"/mnt/work_3/sijin/CAI/sequence.txt\", \"fasta\"):\n",
    "    nm_id = rec.description.split(\"_cds\")[0].split(\"|\")[-1]\n",
    "    seq = str(rec.seq).upper()\n",
    "    records.append({\"NM_ID\": nm_id, \"CDS\": seq, \"CDS_len\": len(seq)})\n",
    "\n",
    "allprot_df = pd.DataFrame(records)\n",
    "\n",
    "all_lengths = pd.DataFrame({\n",
    "    \"Length\": pd.concat([\n",
    "        codon97_df[\"CDS_len\"],\n",
    "        rna_df[\"CDS_len\"],\n",
    "        protein_df[\"CDS_len\"],\n",
    "        allprot_df[\"CDS_len\"]\n",
    "    ], ignore_index=True),\n",
    "    \"Source\": (\n",
    "        [\"Codon97\"]   * len(codon97_df) +\n",
    "        [\"RNAseq\"]    * len(rna_df) +\n",
    "        [\"Proteomics\"]* len(protein_df) +\n",
    "        [\"AllProt\"]   * len(allprot_df)\n",
    "    )\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0a764ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# After sort the dataset, leave the outliers\n",
    "\n",
    "allprot_trimmed = allprot_df.sort_values(\"CDS_len\", ascending=False).iloc[2:]\n",
    "rna_trimmed     = rna_df.sort_values(\"CDS_len\", ascending=False).iloc[1:]\n",
    "\n",
    "all_lengths_trimmed = pd.DataFrame({\n",
    "    \"Length\": pd.concat([\n",
    "        codon97_df[\"CDS_len\"],\n",
    "        rna_trimmed[\"CDS_len\"],\n",
    "        protein_df[\"CDS_len\"],\n",
    "        allprot_trimmed[\"CDS_len\"]\n",
    "    ], ignore_index=True),\n",
    "    \"Source\": (\n",
    "        [\"Codon97\"]   * len(codon97_df) +\n",
    "        [\"RNAseq_Log2FC>0\"]    * len(rna_trimmed) +\n",
    "        [\"Proteomics_Log2FC>0\"]* len(protein_df) +\n",
    "        [\"All_otherprot_all_Log2FC\"]   * len(allprot_trimmed)\n",
    "    )\n",
    "})\n",
    "\n",
    "custom_palette = {\n",
    "    \"Codon97\": \"#8DD3C7\", \n",
    "    \"Proteomics_Log2FC>0\": \"#f6f6bc\",               \n",
    "    \"RNAseq_Log2FC>0\": \"#c2bed6\",\n",
    "    \"All_otherprot_all_Log2FC\": \"#eab375\"  \n",
    "}\n",
    "\n",
    "order = [\"Codon97\", \"Proteomics_Log2FC>0\", \"RNAseq_Log2FC>0\", \"All_otherprot_all_Log2FC\"]\n",
    "counts = all_lengths_trimmed.groupby(\"Source\").size()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21c302cb",
   "metadata": {},
   "source": [
    "#### bootstrap test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55d0aacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "def bootstrap_pvalue(test_vals, all_vals, n_iter=1000, seed=123):\n",
    "    np.random.seed(seed)\n",
    "    test_mean = np.mean(test_vals)\n",
    "    n = len(test_vals)\n",
    "\n",
    "    boot_means = []\n",
    "    for _ in range(n_iter):\n",
    "        sample = np.random.choice(all_vals, size=n, replace=False)\n",
    "        boot_means.append(np.mean(sample))\n",
    "\n",
    "    boot_means = np.array(boot_means)\n",
    "    pval = np.mean(np.abs(boot_means - np.mean(all_vals)) >= \n",
    "                   np.abs(test_mean - np.mean(all_vals)))\n",
    "\n",
    "    return test_mean, np.mean(all_vals), pval\n",
    "\n",
    "\n",
    "def bootstrap_pvalue_median(test_vals, all_vals, n_iter=1000, seed=123):\n",
    "    np.random.seed(seed)\n",
    "    test_median = np.median(test_vals)\n",
    "    n = len(test_vals)\n",
    "\n",
    "    boot_medians = []\n",
    "    for _ in range(n_iter):\n",
    "        sample = np.random.choice(all_vals, size=n, replace=False)\n",
    "        boot_medians.append(np.median(sample))\n",
    "\n",
    "    boot_medians = np.array(boot_medians)\n",
    "    pval = np.mean(np.abs(boot_medians - np.median(all_vals)) >= \n",
    "                   np.abs(test_median - np.median(all_vals)))\n",
    "\n",
    "    return test_median, np.median(all_vals), pval\n",
    "\n",
    "allprot_trimmed = allprot_df.sort_values(\"CDS_len\", ascending=False).iloc[2:]\n",
    "rna_trimmed     = rna_df.sort_values(\"CDS_len\", ascending=False).iloc[1:]\n",
    "\n",
    "codon97_len = codon97_df[\"CDS_len\"].dropna().values\n",
    "rna_len     = rna_trimmed[\"CDS_len\"].dropna().values\n",
    "protein_len = protein_df[\"CDS_len\"].dropna().values\n",
    "all_len     = allprot_trimmed[\"CDS_len\"].dropna().values   \n",
    "\n",
    "# mean\n",
    "results_mean = {}\n",
    "for label, values in {\n",
    "    \"Codon97\": codon97_len,\n",
    "    \"RNAseq\": rna_len,\n",
    "    \"Proteomics\": protein_len\n",
    "}.items():\n",
    "    results_mean[label] = bootstrap_pvalue(values, all_len, n_iter=1000)\n",
    "\n",
    "res_df_mean = pd.DataFrame(results_mean, index=[\"Mean_Group\",\"Mean_Allprot\",\"p-value\"]).T\n",
    "print(\"=== Mean comparison (trimmed) ===\")\n",
    "print(res_df_mean)\n",
    "\n",
    "# median\n",
    "results_median = {}\n",
    "for label, values in {\n",
    "    \"Codon97\": codon97_len,\n",
    "    \"RNAseq\": rna_len,\n",
    "    \"Proteomics\": protein_len\n",
    "}.items():\n",
    "    results_median[label] = bootstrap_pvalue_median(values, all_len, n_iter=1000)\n",
    "\n",
    "res_df_median = pd.DataFrame(results_median, index=[\"Median_Group\",\"Median_Allprot\",\"p-value\"]).T\n",
    "print(\"\\n=== Median comparison (trimmed) ===\")\n",
    "print(res_df_median)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8090d204",
   "metadata": {},
   "source": [
    "#### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e65c6c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "\n",
    "plt.figure(figsize=(10,6))\n",
    "ax = sns.violinplot(\n",
    "    x=\"Source\", y=\"Length\", log_scale=True, data=all_lengths_trimmed,\n",
    "    inner=None, palette=custom_palette, order=order, linewidth=1.2\n",
    ")\n",
    "\n",
    "sns.stripplot(\n",
    "    x=\"Source\", y=\"Length\", data=all_lengths_trimmed,\n",
    "    color=\"black\", alpha=0.2, jitter=0.25, size=1.5, order=order\n",
    ")\n",
    "\n",
    "sns.boxplot(\n",
    "    x=\"Source\", y=\"Length\", data=all_lengths_trimmed,\n",
    "    width=0.15, showcaps=True, order=order,\n",
    "    boxprops={'facecolor':'grey', 'edgecolor':'black'},\n",
    "    whiskerprops={'color':'black'},\n",
    "    medianprops={'color':'black', 'linewidth':1},\n",
    "    showfliers=False\n",
    ")\n",
    "\n",
    "group_means = all_lengths_trimmed.groupby(\"Source\")[\"Length\"].mean()\n",
    "for i, src in enumerate(order):\n",
    "    ax.scatter(i, group_means[src], color=\"white\", zorder=3, s=30, edgecolors=\"black\")\n",
    "    ax.text(i, group_means[src]*1.05, f\"mean = {group_means[src]:.0f}\", \n",
    "            ha='center', va='bottom', fontsize=10, color=\"blue\")\n",
    "\n",
    "p_values = {\n",
    "    \"Codon97\": 0.659,\n",
    "    \"RNAseq_Log2FC>0\": 0.307,\n",
    "    \"Proteomics_Log2FC>0\": 0.000\n",
    "}\n",
    "\n",
    "y_max = all_lengths_trimmed[\"Length\"].max()\n",
    "base = math.log10(y_max)\n",
    "\n",
    "log_step = 0.4  \n",
    "\n",
    "for i, src in enumerate(order[:-1]):  \n",
    "    x1, x2 = i, len(order)-1\n",
    "    pval = p_values.get(src, None)\n",
    "    if pval is None:\n",
    "        continue\n",
    "    \n",
    "    y = 10**(base + (i+1)*log_step)\n",
    "    line_height = 10**(base + (i+1)*log_step + 0.05)\n",
    "    \n",
    "    plt.plot([x1, x1, x2, x2], [y, line_height, line_height, y],\n",
    "             lw=1.2, c='k')\n",
    "    plt.text((x1+x2)/2, line_height*1.1, f\"p = {pval:.3f}\",\n",
    "             ha='center', va='bottom', fontsize=10, color=\"blue\")\n",
    "\n",
    "plt.xticks(\n",
    "    ticks=range(len(order)),\n",
    "    labels=[f\"{src}\\n(n={counts[src]})\" for src in order]\n",
    ")\n",
    "\n",
    "y_min, y_max = ax.get_ylim()\n",
    "ax.set_ylim(0, 1e6) \n",
    "plt.title(\"CDS Length Distribution (outliers removed, log10 scale)\", fontsize=14)\n",
    "plt.ylabel(\"CDS Length (nt, log10)\", fontsize=12)\n",
    "plt.xlabel(\"\")\n",
    "plt.grid(axis=\"y\", linestyle=\"--\", alpha=0.6)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
