{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc1177e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e66d4ee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_counts_file = \"data/Proteomics Counts.csv\"     \n",
    "rna_file = \"data/RNA Seq Stats.csv\"            \n",
    "prot_stats_file = \"data/Proteomics Stats.csv\"      \n",
    "\n",
    "prot_counts = pd.read_csv(prot_counts_file)   \n",
    "rna_df = pd.read_csv(rna_file)                \n",
    "prot_stats = pd.read_csv(prot_stats_file)     "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57fb663b",
   "metadata": {},
   "source": [
    "### STEP1: Find the UniProt for RNA seq and the Proteomics file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b41b9118",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import mygene\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "\n",
    "# Clean the ENSG（.1, .2…）\n",
    "ensg_ids = rna_df[\"Unnamed: 0\"].str.replace(r\"\\.\\d+$\", \"\", regex=True).tolist()\n",
    "\n",
    "# Use ENSG to find UniProt + GeneSymbol\n",
    "rna_map = mg.querymany(\n",
    "    ensg_ids,\n",
    "    scopes=\"ensembl.gene\",\n",
    "    fields=\"symbol,uniprot.Swiss-Prot\",\n",
    "    species=\"human\"\n",
    ")\n",
    "\n",
    "rna_map_df = pd.DataFrame(rna_map)\n",
    "\n",
    "rna_map_df[\"UniProt\"] = rna_map_df[\"uniprot\"].apply(\n",
    "    lambda x: x.get(\"Swiss-Prot\") if isinstance(x, dict) else x\n",
    "    if isinstance(x, str) else None\n",
    ")\n",
    "\n",
    "rna_map_df = rna_map_df[[\"query\",\"symbol\",\"UniProt\"]].rename(\n",
    "    columns={\"query\":\"ENSG\",\"symbol\":\"GeneSymbol\"}\n",
    ")\n",
    "\n",
    "# Merge the table\n",
    "rna_df[\"ENSG_clean\"] = rna_df[\"Unnamed: 0\"].str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
    "rna_df = rna_df.merge(rna_map_df, left_on=\"ENSG_clean\", right_on=\"ENSG\", how=\"left\")\n",
    "\n",
    "rna_df = rna_df.drop(columns=[\"ENSG\",\"ENSG_clean\"])\n",
    "\n",
    "rna_df.to_csv(\"data/RNA_Seq_with_Symbol.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541a3eed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use proteomics_counts to find the Uniprot\n",
    "prot_map = prot_counts[[\"PG.ProteinGroups\", \"PG.ProteinNames\"]].rename(\n",
    "    columns={\"PG.ProteinGroups\":\"UniProt\", \"PG.ProteinNames\":\"EntryName\"}\n",
    ").drop_duplicates()\n",
    "\n",
    "# Prot Stats + Prot Counts merge by Entryname\n",
    "prot_stats = prot_stats.rename(columns={\"G.ProteinNames\":\"EntryName\"})\n",
    "prot_stats_merged = prot_stats.merge(prot_map, on=\"EntryName\", how=\"left\")\n",
    "\n",
    "out_file = \"data/Proteomics_Stats_with_UniProt.csv\"\n",
    "prot_stats_merged.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17394416",
   "metadata": {},
   "source": [
    "### STEP2: Merged RNA seq and Proteomics File using the UniProt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5fd1b7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_file = pd.read_csv(\"data/RNA_Seq_with_Symbol.csv\")\n",
    "prot_file = pd.read_csv(\"data/Proteomics_Stats_with_UniProt.csv\")\n",
    "\n",
    "# Merged two files based on the Uniprot\n",
    "merged = prot_df.merge(rna_df, on=\"UniProt\", how=\"inner\")\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_Merged.csv\"\n",
    "merged.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "745fcb06",
   "metadata": {},
   "source": [
    "### STEP3: Filter the merged file with log2fc>0 and p/q value <0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5381cbcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_file = \"data/RNA_Proteomics_Merged.csv\"\n",
    "merged = pd.read_csv(merged_file)\n",
    "\n",
    "# Apply the filters\n",
    "filtered = merged[\n",
    "    (merged[\"log2FoldChange\"] > 0) &\n",
    "    (merged[\"Log2FC\"] > 0) &\n",
    "    (merged[\"Welch's T-test q-value ZFC3H1 KO\"] < 0.05) &\n",
    "    (merged[\"padj\"] < 0.05)\n",
    "]\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_Filtered.csv\"\n",
    "filtered.to_csv(out_file, index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b4e3fd4",
   "metadata": {},
   "source": [
    "#### Find NM_id and use that to find CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f8e1ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "rna_df = pd.read_csv(\"data/RNA_Proteomics_Filtered.csv\")\n",
    "rna_df = rna_df.rename(columns={\"Unnamed: 0\": \"ENSG\"})\n",
    "\n",
    "mane_df = pd.read_csv(\"MANE.GRCh38.v1.4.summary.txt\", sep=\"\\t\", comment = None)\n",
    "\n",
    "mane_select = mane_df[mane_df[\"MANE_status\"] == \"MANE Select\"]\n",
    "\n",
    "ensg_to_nm = dict(zip(mane_select[\"symbol\"], mane_select[\"RefSeq_nuc\"]))\n",
    "\n",
    "rna_df[\"NM_ID\"] = rna_df[\"GeneSymbol\"].map(ensg_to_nm)\n",
    "\n",
    "rna_df.to_csv(\"data/RNA_Proteomics_Filtered.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "262f8604",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "Entrez.email = \"sijin@ad.unc.edu\"\n",
    "\n",
    "def fetch_cds(nm_id):\n",
    "    \"\"\"Fetch CDS nucleotide sequence for a given NM accession\"\"\"\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=nm_id, rettype=\"gb\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"genbank\")\n",
    "        handle.close()\n",
    "\n",
    "        # Find CDS feature\n",
    "        for feature in record.features:\n",
    "            if feature.type == \"CDS\":\n",
    "                seq = feature.extract(record.seq)\n",
    "                return str(seq)\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"Failed for {nm_id}: {e}\")\n",
    "        return None\n",
    "\n",
    "# Fetch CDS for each NM_ID\n",
    "cds_dict = {}\n",
    "for nm in rna_df[\"NM_ID\"].dropna():\n",
    "    cds_dict[nm] = fetch_cds(nm)\n",
    "\n",
    "# Add column back\n",
    "rna_df[\"CDS\"] = rna_df[\"NM_ID\"].map(cds_dict)\n",
    "\n",
    "# Save\n",
    "rna_df.to_csv(\"data/RNA_Proteomics_Filtered.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72a97957",
   "metadata": {},
   "source": [
    "### STEP4: RNA targets not affected at protein level (log2FC>0 RNA-seq and padj<0.05 RNA-seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8724fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply filters\n",
    "filtered = merged[\n",
    "    (merged[\"log2FoldChange\"] > 0) &\n",
    "    (merged[\"padj\"] < 0.05)\n",
    "]\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_RNAlevel.csv\"\n",
    "filtered.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4df3e634",
   "metadata": {},
   "source": [
    "#### Fine CDS of the RNA Level file using NM_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eddf99d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "rna_seq_df = pd.read_csv(\"data/RNA_Proteomics_RNAlevel.csv\")\n",
    "rna_seq_df = rna_seq_df.rename(columns={\"Unnamed: 0\": \"ENSG\"})\n",
    "\n",
    "# Load MANE summary file-download from NCBI website\n",
    "mane_df = pd.read_csv(\"MANE.GRCh38.v1.4.summary.txt\", sep=\"\\t\", comment = None)\n",
    "\n",
    "# Keep only MANE Select rows\n",
    "mane_select = mane_df[mane_df[\"MANE_status\"] == \"MANE Select\"]\n",
    "\n",
    "# Build GeneSymbol→ NM mapping\n",
    "ensg_to_nm = dict(zip(mane_select[\"symbol\"], mane_select[\"RefSeq_nuc\"]))\n",
    "\n",
    "# Map NM_ID using your 'GeneSymbol' column\n",
    "rna_seq_df[\"NM_ID\"] = rna_seq_df[\"GeneSymbol\"].map(ensg_to_nm)\n",
    "\n",
    "# Save back to the same file\n",
    "rna_seq_df.to_csv(\"data/RNA_Proteomics_RNAlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b42d7190",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "Entrez.email = \"sijin@ad.unc.edu\"\n",
    "\n",
    "rna_seq_df[\"CDS\"] = rna_seq_df[\"NM_ID\"].map(cds_dict)\n",
    "\n",
    "# Save\n",
    "rna_seq_df.to_csv(\"data/RNA_Proteomics_RNAlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6db209c5",
   "metadata": {},
   "source": [
    "#### Find the CAI value for each gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "598326ec",
   "metadata": {},
   "source": [
    "### STEP5: RNA targets affected at protein level (log2FC>0 RNA and protein and padj<0.05)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b681e6e",
   "metadata": {},
   "source": [
    "#### 5.1 Find NM_id; 5.2 Find CDS use NM_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44123ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply filters\n",
    "filtered = merged[\n",
    "    (merged[\"Log2FC\"] > 0) &\n",
    "    (merged[\"Welch's T-test q-value ZFC3H1 KO\"] < 0.05)\n",
    "]\n",
    "\n",
    "out_file = \"data/RNA_Proteomics_Proteinlevel.csv\"\n",
    "filtered.to_csv(out_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92687ece",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot_seq_df = pd.read_csv(\"data/RNA_Proteomics_Proteinlevel.csv\")\n",
    "prot_seq_df = prot_seq_df.rename(columns={\"Unnamed: 0\": \"ENSG\"})\n",
    "\n",
    "mane_df = pd.read_csv(\"MANE.GRCh38.v1.4.summary.txt\", sep=\"\\t\", comment = None)\n",
    "\n",
    "mane_select = mane_df[mane_df[\"MANE_status\"] == \"MANE Select\"]\n",
    "\n",
    "ensg_to_nm = dict(zip(mane_select[\"symbol\"], mane_select[\"RefSeq_nuc\"]))\n",
    "\n",
    "prot_seq_df[\"NM_ID\"] = prot_seq_df[\"GeneSymbol\"].map(ensg_to_nm)\n",
    "\n",
    "prot_seq_df.to_csv(\"data/RNA_Proteomics_Proteinlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f7720c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use nm_id to find CDS\n",
    "prot_seq_df[\"CDS\"] = prot_seq_df[\"NM_ID\"].map(cds_dict)\n",
    "prot_seq_df.to_csv(\"data/RNA_Proteomics_Proteinlevel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f44f56",
   "metadata": {},
   "source": [
    "### STEP6: All protein's distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "129c3ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import mygene\n",
    "\n",
    "# Input files\n",
    "prot_counts_file = \"data/Proteomics Counts.csv\"\n",
    "prot_stats_file  = \"data/Proteomics Stats.csv\"\n",
    "prot_counts = pd.read_csv(prot_counts_file)\n",
    "prot_stats  = pd.read_csv(prot_stats_file)\n",
    "\n",
    "# Map the Uniprot\n",
    "prot_map = prot_counts[[\"PG.ProteinGroups\", \"PG.ProteinNames\"]].rename(\n",
    "    columns={\"PG.ProteinGroups\":\"UniProt\", \"PG.ProteinNames\":\"EntryName\"}\n",
    ").drop_duplicates()\n",
    "\n",
    "# Merge the file\n",
    "prot_stats = prot_stats.rename(columns={\"G.ProteinNames\":\"EntryName\"})\n",
    "prot_stats_merged = prot_stats.merge(prot_map, on=\"EntryName\", how=\"left\")\n",
    "\n",
    "# Find the Genesymbol by uniprot \n",
    "mg = mygene.MyGeneInfo()\n",
    "res = mg.querymany(\n",
    "    prot_stats_merged[\"UniProt\"].dropna().tolist(),\n",
    "    scopes=\"uniprot\",\n",
    "    fields=\"symbol\",\n",
    "    species=\"human\"\n",
    ")\n",
    "\n",
    "map_df = pd.DataFrame(res)[[\"query\",\"symbol\"]].rename(\n",
    "    columns={\"query\":\"UniProt\",\"symbol\":\"GeneSymbol\"}\n",
    ")\n",
    "\n",
    "prot_stats_final = prot_stats_merged.merge(map_df, on=\"UniProt\", how=\"left\")\n",
    "\n",
    "out_file = \"data/All_Prot_Level.csv\"\n",
    "prot_stats_final.to_csv(out_file, index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fa3fab8",
   "metadata": {},
   "source": [
    "#### Find nm_id and use nm_id to find CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44fb2dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df = pd.read_csv(\"data/All_Prot_Level.csv\")\n",
    "mane_df = pd.read_csv(\"MANE.GRCh38.v1.4.summary.txt\", sep=\"\\t\", comment = None)\n",
    "mane_select = mane_df[mane_df[\"MANE_status\"] == \"MANE Select\"]\n",
    "ensg_to_nm = dict(zip(mane_select[\"symbol\"], mane_select[\"RefSeq_nuc\"]))\n",
    "\n",
    "all_df[\"NM_ID\"] = all_df[\"GeneSymbol\"].map(ensg_to_nm)\n",
    "\n",
    "all_df.to_csv(\"data/All_Prot_Level.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c9a9329",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df[\"NM_ID\"].dropna().drop_duplicates().to_csv(\"data/my_ids.txt\", index=False, header=False)\n",
    "\n",
    "#then upload the nm_id's to NCBI and download the CDS in sequence.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
